File: src/test/java/org/oewntk/model/TestKeys.java
/*
 * Copyright (c) 2021. Bernard Bou.
 */

package org.oewntk.model;

import org.junit.Test;

import java.util.Set;

import static org.junit.Assert.*;

public class TestKeys
{
	@Test
	public void testRow()
	{
		String wordRow = "row";
		Pronunciation pRowOu = new Pronunciation("ɹəʊ", null);
		Pronunciation pRowAu = new Pronunciation("ɹaʊ", null);

		Lex lexRowOu = new Lex(wordRow, "n-1", "source1").setPronunciations(pRowOu);
		Lex lexRowAu = new Lex(wordRow, "n-2", "source2").setPronunciations(pRowAu);
		Lex lexRowOuN = new Lex(wordRow, "n", "source1").setPronunciations(pRowOu);
		Lex lexRowAuN = new Lex(wordRow, "n", "source2").setPronunciations(pRowAu);

		assertNotEquals(Key.W_P_A.of_t(lexRowOu), Key.W_P_A.of_t(lexRowAu));
		assertNotEquals(Key.W_P_D.of_t(lexRowOu), Key.W_P_D.of_t(lexRowAu)); // because discriminant is different
		assertEquals(Key.W_P_D.of_t(lexRowOuN), Key.W_P_D.of_t(lexRowAuN)); // because discriminant is same
		assertEquals(Key.W_P.of(lexRowOu, Lex::getLemma, Lex::getType), Key.W_P.of(lexRowAu, Lex::getLemma, Lex::getType));
	}

	@Test
	public void testMobile()
	{
		String wordMobile = "mobile";
		Pronunciation pMobile = new Pronunciation("ˈməʊbaɪl", null);
		Pronunciation pMobileGB = new Pronunciation("ˈməʊbaɪl", "GB");
		Pronunciation pMobileUS = new Pronunciation("ˈmoʊbil", "US");
		Pronunciation[] paMobile1 = new Pronunciation[]{pMobile, pMobileGB, pMobileUS};
		Pronunciation[] paMobile2 = new Pronunciation[]{pMobile, pMobileGB, pMobileUS};

		Lex lexMobile0 = new Lex(wordMobile, "n", "source1");
		Lex lexMobile1 = new Lex(wordMobile, "n", "source1").setPronunciations(paMobile1);
		Lex lexMobile2 = new Lex(wordMobile, "n", "source2").setPronunciations(paMobile2);

		assertEquals(Key.W_P_A.of_t(lexMobile1), Key.W_P_A.of_t(lexMobile2));
		assertEquals(Key.W_P_D.of_t(lexMobile1), Key.W_P_D.of_t(lexMobile2));
		assertEquals(Key.W_P.of_t(lexMobile1), Key.W_P.of_t(lexMobile2));

		assertNotEquals(Key.W_P_A.of_t(lexMobile1), Key.W_P_A.of_t(lexMobile0));
		assertEquals(Key.W_P_D.of_t(lexMobile1), Key.W_P_D.of_t(lexMobile0)); // because discriminant is both null
		assertEquals(Key.W_P.of_t(lexMobile1), Key.W_P.of_t(lexMobile0));
	}

	@Test
	public void testCritical()
	{
		String wordCritical = "critical";
		Lex lexCriticalA = new Lex(wordCritical, "a", "source1");
		Lex lexCriticalS = new Lex(wordCritical, "s", "source2");

		assertNotEquals(Key.W_P_A.of_t(lexCriticalA), Key.W_P_A.of_t(lexCriticalS));
		assertNotEquals(Key.W_P_D.of_t(lexCriticalA), Key.W_P_D.of_t(lexCriticalS));
		assertEquals(Key.W_P.of_p(lexCriticalA), Key.W_P.of_p(lexCriticalS));
		assertEquals(Key.W_P_A.of_p(lexCriticalA), Key.W_P_A.of_p(lexCriticalS)); // A and S are merged
	}

	@Test
	public void testCapitalisation()
	{
		String wordEarthL = "earth";
		String wordEarthU = "Earth";
		Lex lexEarthL = new Lex(wordEarthL, "n", "source1");
		Lex lexEarthU = new Lex(wordEarthU, "n", "source2");

		assertNotEquals(Key.W_P_A.of_t(lexEarthL), Key.W_P_A.of_t(lexEarthU));
		assertNotEquals(Key.W_P_D.of_t(lexEarthL), Key.W_P_D.of_t(lexEarthU));
		assertNotEquals(Key.W_P.of_lc_t(lexEarthL), Key.W_P.of_lc_t(lexEarthU));
		assertEquals(Key.W_P_A.of_lc_t(lexEarthL), Key.W_P_A.of_lc_t(lexEarthU));
	}

	@Test
	public void testPronunciations()
	{
		Pronunciation pMobile = new Pronunciation("ˈməʊbaɪl", null);
		Pronunciation pMobileGB = new Pronunciation("ˈməʊbaɪl", "GB");
		Pronunciation pMobileUS = new Pronunciation("ˈmoʊbil", "US");
		Pronunciation[] paMobile1 = new Pronunciation[]{pMobile, pMobileGB, pMobileUS};
		Pronunciation[] paMobile2 = new Pronunciation[]{pMobile, pMobileGB, pMobileUS};
		Set<Pronunciation> psMobile1 = Set.of(paMobile1);
		Set<Pronunciation> psMobile2 = Set.of(paMobile2);

		assertEquals(pMobile, new Pronunciation("ˈməʊbaɪl", null));
		assertEquals(pMobileGB, new Pronunciation("ˈməʊbaɪl", "GB"));
		assertNotEquals(pMobile, null);
		assertNotEquals(pMobile, pMobileGB);
		assertNotEquals(pMobile, pMobileUS);
		assertNotEquals(pMobileGB, pMobileUS);

		assertArrayNotEquals(paMobile1, null);
		assertArrayNotEquals(paMobile1, new Pronunciation[0]);
		assertArrayNotEquals(paMobile1, new Pronunciation[]{pMobileGB, pMobile, pMobileUS});
		assertArrayNotEquals(paMobile2, new Pronunciation[]{pMobileGB, pMobileUS, pMobile});
		assertArrayEquals(paMobile1, new Pronunciation[]{pMobile, pMobileGB, pMobileUS});
		assertArrayEquals(paMobile2, new Pronunciation[]{pMobile, pMobileGB, pMobileUS});

		assertEquals(psMobile1, Set.of(paMobile1));
		assertEquals(psMobile2, Set.of(paMobile2));
		assertEquals(psMobile1, Set.of(new Pronunciation[]{pMobileGB, pMobile, pMobileUS}));
		assertEquals(psMobile2, Set.of(new Pronunciation[]{pMobile, pMobileGB, pMobileUS}));
		assertEquals(psMobile1, Set.of(new Pronunciation[]{new Pronunciation("ˈməʊbaɪl", "GB"), new Pronunciation("ˈməʊbaɪl", null), new Pronunciation("ˈmoʊbil", "US")}));
		assertEquals(psMobile2, Set.of(new Pronunciation[]{new Pronunciation("ˈməʊbaɪl", null), new Pronunciation("ˈməʊbaɪl", "GB"), new Pronunciation("ˈmoʊbil", "US")}));

		assertNotEquals(paMobile1, paMobile2);
		assertEquals(psMobile1, psMobile2);

		Pronunciation pRowOu = new Pronunciation("ɹəʊ", null);
		Pronunciation pRowAu = new Pronunciation("ɹaʊ", null);
		Pronunciation[] paRow1 = new Pronunciation[]{pRowOu, pRowAu};
		Pronunciation[] paRow2 = new Pronunciation[]{pRowAu, pRowOu};
		Set<Pronunciation> psRow1 = Set.of(paRow1);
		Set<Pronunciation> psRow2 = Set.of(paRow2);

		assertNotEquals(pRowOu, pRowAu);
		assertNotEquals(paRow1, paRow2);
		assertEquals(psRow1, psRow2);
	}

	private void assertArrayNotEquals(final Object[] expecteds, final Object[] actuals)
	{
		try
		{
			assertArrayEquals(expecteds, actuals);
		}
		catch (AssertionError e)
		{
			return;
		}
		fail();
	}
}

File: src/test/java/org/oewntk/model/TestKeys.kt
/*
 * Copyright (c) 2021-2024. Bernard Bou.
 */
package org.oewntk.model

import org.junit.Assert.*
import org.junit.Test
import java.io.PrintStream
import kotlin.test.assertContentEquals

class TestKeys {

    @Test
    fun testRow() {
        val wordRow = "row"
        val pRowOu = Pronunciation(PronunciationValue("ɹəʊ"), null)
        val pRowAu = Pronunciation(PronunciationValue("ɹaʊ"), null)

        val lexRowOu = Lex(Lemma(wordRow), "n-1").apply { pronunciations = setOf(pRowOu) }
        val lexRowAu = Lex(Lemma(wordRow), "n-2").apply { pronunciations = setOf(pRowAu) }
        val lexRowOuN = Lex(Lemma(wordRow), "n").apply { pronunciations = setOf(pRowOu) }
        val lexRowAuN = Lex(Lemma(wordRow), "n").apply { pronunciations = setOf(pRowAu) }

        assertNotEquals(Key.UsingPronunciation.of(lexRowOu), Key.UsingPronunciation.of(lexRowAu))
        assertNotEquals(Key.UsingDiscriminant.of(lexRowOu), Key.UsingDiscriminant.of(lexRowAu)) // because discriminant is different
        assertEquals(Key.UsingDiscriminant.of(lexRowOuN), Key.UsingDiscriminant.of(lexRowAuN)) // because discriminant is same
        assertEquals(
            Key.Base.of(lexRowOu, Lex::lemma) { it.partOfSpeech.toCategory() },
            Key.Base.of(lexRowAu, Lex::lemma) { it.partOfSpeech.toCategory() }
        )
    }

    @Test
    fun testEquals() {
        val wordMobile = "mobile"
        val pMobile = Pronunciation(PronunciationValue("ˈməʊbaɪl"), null)
        val pMobileGB = Pronunciation(PronunciationValue("ˈməʊbaɪl"), "GB")
        val pMobileUS = Pronunciation(PronunciationValue("ˈmoʊbil"), "US")
        val paMobile1 = arrayOf(pMobile, pMobileGB, pMobileUS)
        val paMobile2 = arrayOf(pMobile, pMobileUS, pMobileGB)
        ps.println("pronunciations1 = ${paMobile1.withIndex().joinToString(separator = " ") { "#${it.index} ${it.value}" }}")
        ps.println("pronunciations2 = ${paMobile2.withIndex().joinToString(separator = " ") { "#${it.index} ${it.value}" }}")
        assertFalse(paMobile1.contentEquals(paMobile2))
        assertFalse(paMobile1.contentEquals(paMobile2))
        assertEquals(paMobile1.toSet(), paMobile2.toSet())

        val lexMobile0 = Lex(Lemma(wordMobile), "n")
        val lexMobile1 = Lex(Lemma(wordMobile), "n").apply { pronunciations = paMobile1.toSet() }
        val lexMobile2 = Lex(Lemma(wordMobile), "n").apply { pronunciations = paMobile2.toSet() }
        ps.println("lex0 = $lexMobile0")
        ps.println("lex1 = $lexMobile1")
        ps.println("lex2 = $lexMobile2")

        val k0 = Key.UsingPronunciation.of(lexMobile0)
        val k1 = Key.UsingPronunciation.of(lexMobile1)
        val k2 = Key.UsingPronunciation.of(lexMobile2)
        ps.println("key0 = $k0")
        ps.println("key1 = $k1")
        ps.println("key2 = $k2")

        val p0 = k0.pronunciations ?: emptySet()
        val p1 = k1.pronunciations ?: emptySet()
        val p2 = k2.pronunciations ?: emptySet()
        assertEquals(p1, p2)
        assertNotEquals(p1, p0)

        assertEquals(k1, k2)
        assertNotEquals(k1, k0)
    }

    @Test
    fun testMobile() {
        val wordMobile = "mobile"
        val pMobile = Pronunciation(PronunciationValue("ˈməʊbaɪl"), null)
        val pMobileGB = Pronunciation(PronunciationValue("ˈməʊbaɪl"), "GB")
        val pMobileUS = Pronunciation(PronunciationValue("ˈmoʊbil"), "US")
        val paMobile1 = arrayOf(pMobile, pMobileGB, pMobileUS)
        val paMobile2 = arrayOf(pMobile, pMobileGB, pMobileUS)

        val lexMobile0 = Lex(Lemma(wordMobile), "n").apply { pronunciations = null }
        val lexMobile1 = Lex(Lemma(wordMobile), "n").apply { pronunciations = paMobile1.toSet() }
        val lexMobile2 = Lex(Lemma(wordMobile), "n").apply { pronunciations = paMobile2.toSet() }

        assertEquals(Key.UsingPronunciation.of(lexMobile1), Key.UsingPronunciation.of(lexMobile2))
        assertEquals(Key.UsingDiscriminant.of(lexMobile1), Key.UsingDiscriminant.of(lexMobile2))
        assertEquals(Key.Base.of(lexMobile1), Key.Base.of(lexMobile2))

        assertNotEquals(Key.UsingPronunciation.of(lexMobile1), Key.UsingPronunciation.of(lexMobile0))
        assertEquals(Key.UsingDiscriminant.of(lexMobile1), Key.UsingDiscriminant.of(lexMobile0)) // because discriminant is both null
        assertEquals(Key.Base.of(lexMobile1), Key.Base.of(lexMobile0))
    }

    @Test
    fun testCapitalisation() {
        val wordEarthL = "earth"
        val wordEarthU = "Earth"
        val lexEarthL = Lex(Lemma(wordEarthL), "n")
        val lexEarthU = Lex(Lemma(wordEarthU), "n")

        assertNotEquals(Key.UsingPronunciation.of(lexEarthL), Key.UsingPronunciation.of(lexEarthU))
        assertNotEquals(Key.UsingDiscriminant.of(lexEarthL), Key.UsingDiscriminant.of(lexEarthU))
        assertEquals(Key.Base.ofIgnoringCase(lexEarthL), Key.Base.ofIgnoringCase(lexEarthU))
        assertEquals(Key.UsingPronunciation.ofIgnoringCase(lexEarthL), Key.UsingPronunciation.ofIgnoringCase(lexEarthU))
    }

    @Test
    fun testPronunciations() {
        val pMobile = Pronunciation(PronunciationValue("ˈməʊbaɪl"), null)
        val pMobileGB = Pronunciation(PronunciationValue("ˈməʊbaɪl"), "GB")
        val pMobileUS = Pronunciation(PronunciationValue("ˈmoʊbil"), "US")
        val paMobile1 = arrayOf(pMobile, pMobileGB, pMobileUS)
        val paMobile2 = arrayOf(pMobile, pMobileGB, pMobileUS)
        val psMobile1 = setOf(*paMobile1)
        val psMobile2 = setOf(*paMobile2)

        assertEquals(pMobile, Pronunciation(PronunciationValue("ˈməʊbaɪl"), null))
        assertEquals(pMobileGB, Pronunciation(PronunciationValue("ˈməʊbaɪl"), "GB"))
        assertNotEquals(pMobile, null)
        assertNotEquals(pMobile, pMobileGB)
        assertNotEquals(pMobile, pMobileUS)
        assertNotEquals(pMobileGB, pMobileUS)

        assertNotEquals(paMobile1, null)
        assertNotEquals(paMobile1, arrayOf<Pronunciation>())
        assertNotEquals(paMobile1, arrayOf(pMobileGB, pMobile, pMobileUS))
        assertNotEquals(paMobile2, arrayOf(pMobileGB, pMobileUS, pMobile))
        assertContentEquals(paMobile1, arrayOf(pMobile, pMobileGB, pMobileUS))
        assertContentEquals(paMobile2, arrayOf(pMobile, pMobileGB, pMobileUS))

        assertEquals(psMobile1, setOf(*paMobile1))
        assertEquals(psMobile2, setOf(*paMobile2))
        assertEquals(psMobile1, setOf(pMobileGB, pMobile, pMobileUS))
        assertEquals(psMobile2, setOf(pMobile, pMobileGB, pMobileUS))
        assertEquals(
            psMobile1,
            setOf(Pronunciation(PronunciationValue("ˈməʊbaɪl"), "GB"), Pronunciation(PronunciationValue("ˈməʊbaɪl"), null), Pronunciation(PronunciationValue("ˈmoʊbil"), "US"))
        )
        assertEquals(
            psMobile2,
            setOf(Pronunciation(PronunciationValue("ˈməʊbaɪl"), null), Pronunciation(PronunciationValue("ˈməʊbaɪl"), "GB"), Pronunciation(PronunciationValue("ˈmoʊbil"), "US"))
        )

        assertNotEquals(paMobile1, paMobile2)
        assertEquals(psMobile1, psMobile2)

        val pRowOu = Pronunciation(PronunciationValue("ɹəʊ"), null)
        val pRowAu = Pronunciation(PronunciationValue("ɹaʊ"), null)
        val paRow1 = arrayOf(pRowOu, pRowAu)
        val paRow2 = arrayOf(pRowAu, pRowOu)
        val psRow1 = setOf(*paRow1)
        val psRow2 = setOf(*paRow2)

        assertNotEquals(pRowOu, pRowAu)
        assertNotEquals(paRow1, paRow2)
        assertEquals(psRow1, psRow2)
    }

    companion object {

        private val silent = !System.getProperties().containsKey("VERBOSE") && if (System.getProperties().containsKey("SILENT")) true
        else true

        private val ps: PrintStream = if (!silent) Tracing.psInfo else Tracing.psNull
    }
}
